### PR TITLE
#4367 Enhance Knapsack problem

### PR DESCRIPTION
**Pull request by Manan-09**
- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] All filenames are in PascalCase.
- [x] All functions and variable names follow Java naming conventions.
- [x] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.

Reduced space complexity of solution from O(n*w) to O(w)
Named variables better
Added testfile

All code is covered in testcases
Screenshot 2023-09-11 at 2 03 35 PM

### CODE REVIEW

File: src/test/java/com/thealgorithms/dynamicprogramming/KnapsackTest.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    @Test
+    public void testKnapSackInvalidInput() {
+        int[] weights = {2, 3, 4};
+        int[] values = {3, 4, 5, 6}; // Different length values array.
+        int weightCapacity = 5;
+        // This test should throw an IllegalArgumentException.
+        assertThrows(IllegalArgumentException.class, () -> { Knapsack.knapSack(weightCapacity, weights, values); });
+    }
+}
```
Comment (vil02) [inline src/test/java/com/thealgorithms/dynamicprogramming/KnapsackTest.java]:
```suggestion
}

```

This will solve the `clang-format` failure.

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java [hunk 2]
```diff
@@ -1,38 +1,40 @@
...
- * A DynamicProgramming based solution for 0-1 Knapsack problem
+ * A Dynamic Programming based solution for the 0-1 Knapsack problem.
+ * This class provides a method, `knapSack`, that calculates the maximum value that can be
+ * obtained from a given set of items with weights and values, while not exceeding a
+ * given weight capacity.
+ *
+ * @see <a href="https://en.wikipedia.org/?title=0-1_Knapsack_problem">0-1 Knapsack Problem </a>
  */
-public class Knapsack {
+public final class Knapsack {
```
Comment (vil02) [inline src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java:11]:
```suggestion
public final class Knapsack {
    private Knapsack() {
    }
```

File: src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java [hunk 3]
```diff
@@ -1,38 +1,40 @@
...
+     * Solves the 0-1 Knapsack problem using Dynamic Programming.
+     *
+     * @param weightCapacity The maximum weight capacity of the knapsack.
+     * @param weights        An array of item weights.
+     * @param values         An array of item values.
+     * @return The maximum value that can be obtained without exceeding the weight capacity.
+     * @throws IllegalArgumentException If the input arrays are null or have different lengths.
+     */
+    public static int knapSack(final int weightCapacity, final int[] weights, final int[] values) throws IllegalArgumentException {
+        if (weights == null || values == null || weights.length != values.length) {
```
Comment (vil02) [inline src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java:23]: Is it really needed to check if `weights` or `values` are `null`?

File: src/test/java/com/thealgorithms/dynamicprogramming/KnapsackTest.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
+package com.thealgorithms.dynamicprogramming;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (vil02) [inline src/test/java/com/thealgorithms/dynamicprogramming/KnapsackTest.java:3]:
```suggestion
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
```

**Review by vil02 — CHANGES_REQUESTED** [review]
There are two other solutions of that problem: [DyanamicProgrammingKnapsack](https://github.com/TheAlgorithms/Java/blob/94621fb63eafcf18666dfad176fffc28f82b439a/src/main/java/com/thealgorithms/dynamicprogramming/DyanamicProgrammingKnapsack.java#L5) and [KnapsackMemoization.java](https://github.com/TheAlgorithms/Java/blob/94621fb63eafcf18666dfad176fffc28f82b439a/src/main/java/com/thealgorithms/dynamicprogramming/KnapsackMemoization.java#L12) - what is the relation between them and the one you update?

Besides that I suggest some _cosmetic_ changes.

**Comment by Manan-09** [comment]
Seems like this is  [DyanamicProgrammingKnapsack](https://github.com/TheAlgorithms/Java/blob/94621fb63eafcf18666dfad176fffc28f82b439a/src/main/java/com/thealgorithms/dynamicprogramming/DyanamicProgrammingKnapsack.java#L5) duplicate of what I just refactored, and  [KnapsackMemoization.java](https://github.com/TheAlgorithms/Java/blob/94621fb63eafcf18666dfad176fffc28f82b439a/src/main/java/com/thealgorithms/dynamicprogramming/KnapsackMemoization.java#L12) is recursive solition of this problem.
But my solution has better space complexity with same TC

**Review by debasishbsws — DISMISSED** [review]
Looks good👍

File: src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java (hunk 2 above)
Comment (vil02) [reply src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java:11]: @Manan-09 why did you mark this as resolved?

File: src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java (hunk 3 above)
Comment (vil02) [reply src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java:25]: @Manan-09 why did you mark this as resolved?

**Comment by vil02** [comment]
> Seems like this is [DyanamicProgrammingKnapsack](https://github.com/TheAlgorithms/Java/blob/94621fb63eafcf18666dfad176fffc28f82b439a/src/main/java/com/thealgorithms/dynamicprogramming/DyanamicProgrammingKnapsack.java#L5) duplicate of what I just refactored, and [KnapsackMemoization.java](https://github.com/TheAlgorithms/Java/blob/94621fb63eafcf18666dfad176fffc28f82b439a/src/main/java/com/thealgorithms/dynamicprogramming/KnapsackMemoization.java#L12) is recursive solition of this problem. But my solution has better space complexity with same TC

@siriak how should we proceed?

**Comment by siriak** [comment]
Please remove the duplicate, but leave the version with memoization as it's a different approach

File: src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java (hunk 3 above)
Comment (siriak) [reply src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java:25]: I think it's a reasonable check. There's nothing else to prevent them from being null

**Review by siriak — APPROVED** [review]

File: src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java (hunk 3 above)
Comment (vil02) [reply src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java:25]: In this case I would highly recommend adding two more tests where the input arrays are `null` (i.e. one with `weights` being `null` and second with `values` being `null`).

File: src/test/java/com/thealgorithms/dynamicprogramming/KnapsackTest.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    @Test
+    public void testKnapSackInvalidInput() {
+        int[] weights = {2, 3, 4};
+        int[] values = {3, 4, 5, 6}; // Different length values array.
+        int weightCapacity = 5;
+        // This test should throw an IllegalArgumentException.
+        assertThrows(IllegalArgumentException.class, () -> { Knapsack.knapSack(weightCapacity, weights, values); });
+    }
```
Comment (vil02) [inline src/test/java/com/thealgorithms/dynamicprogramming/KnapsackTest.java:56]:
```suggestion
    @Test
    public void testKnapSackThrowsForInputsOfDifferentLength() {
        int[] weights = {2, 3, 4};
        int[] values = {3, 4, 5, 6}; // Different length values array.
        int weightCapacity = 5;
        assertThrows(IllegalArgumentException.class, () -> { Knapsack.knapSack(weightCapacity, weights, values); });
    }
```

File: src/test/java/com/thealgorithms/dynamicprogramming/KnapsackTest.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+    @Test
+    public void testKnapSackNullInput() {
+        int[] weights = {2, 3, 4};
+        int[] values = {3, 4, 6};
+        int weightCapacity = 5;
+        // This test should throw an IllegalArgumentException.
+        assertThrows(IllegalArgumentException.class, () -> { Knapsack.knapSack(weightCapacity, null, values); });
+        assertThrows(IllegalArgumentException.class, () -> { Knapsack.knapSack(weightCapacity, weights, null); });
+    }
```
Comment (vil02) [inline src/test/java/com/thealgorithms/dynamicprogramming/KnapsackTest.java:66]:
```suggestion
    @Test
    public void testKnapSackThrowsForNullInputs() {
        int[] weights = {2, 3, 4};
        int[] values = {3, 4, 6};
        int weightCapacity = 5;
        assertThrows(IllegalArgumentException.class, () -> { Knapsack.knapSack(weightCapacity, null, values); });
        assertThrows(IllegalArgumentException.class, () -> { Knapsack.knapSack(weightCapacity, weights, null); });
    }
```

**Review by vil02 — CHANGES_REQUESTED** [review]
I would suggest to use _meaningful_ test names instead of comments. Names will appear in the output in case if they fail, Comments will not.

Thanks for adding missing tests!

Besides that: the code looks good.

File: src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java [hunk 7]
```diff
@@ -1,38 +1,42 @@
...
+    /**
+     * Solves the 0-1 Knapsack problem using Dynamic Programming.
+     *
+     * @param weightCapacity The maximum weight capacity of the knapsack.
+     * @param weights        An array of item weights.
+     * @param values         An array of item values.
+     * @return The maximum value that can be obtained without exceeding the weight capacity.
+     * @throws IllegalArgumentException If the input arrays are null or have different lengths.
+     */
+    public static int knapSack(final int weightCapacity, final int[] weights, final int[] values) throws IllegalArgumentException {
```
Comment (vil02) [inline src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java:24]: What if `weightCapacity` would be negative? Since it is checked if `weights` and `values` are not `null`, I think there should also be a check if provided `weightCapacity` is not negative.

**Review by vil02 — CHANGES_REQUESTED** [review]

**Review by vil02 — CHANGES_REQUESTED** [review]
What if one of the entries of `weights` is negative or `0`?

```java
Knapsack.knapSack(1, new int[]{0, 0, 0, 0}, new int[]{1, 1, 1, 1})
```
gives me `16` and I think the correct answer should be `4`.

**Comment by Manan-09** [comment]
> What if one of the entries of `weights` is negative or `0`?
> 
> The following code:
> 
> ```java
> Knapsack.knapSack(1, new int[]{0, 0, 0, 0}, new int[]{1, 1, 1, 1})
> ```
> 
> gives me `16` and I think the correct answer should be `4`.

Knapsack problem have positive weights normally, should we add extra check for non positive weights ?

**Comment by vil02** [comment]
> Knapsack problem have positive weights normally, should we add extra check for non positive weights ?
Yes! Throwing an error is the best way of saying that non-positive weights are forbidden.

File: src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java [hunk 8]
```diff
@@ -1,38 +1,51 @@
...
+ * @see <a href="https://en.wikipedia.org/?title=0-1_Knapsack_problem">0-1 Knapsack Problem </a>
  */
-public class Knapsack {
+public final class Knapsack {
 
-    private static int knapSack(int W, int[] wt, int[] val, int n) throws IllegalArgumentException {
-        if (wt == null || val == null) {
-            throw new IllegalArgumentException();
+    private Knapsack(){}
+    ;
```
Comment (vil02) [inline src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java:14]: This should solve the `clang-tidy` error.

```suggestion
    private Knapsack(){}
```

File: src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java [hunk 9]
```diff
@@ -1,38 +1,51 @@
...
+            throw new IllegalArgumentException("Weight capacity should not be negative.");
+        }
+        if (weights == null || values == null || weights.length != values.length) {
+            throw new IllegalArgumentException("Input arrays must not be null and must have the same length.");
+        }
+        for (int w : weights) {
+            if (w <= 0) {
+                throw new IllegalArgumentException("Input array should not contain non-positive weight(s).");
+            }
         }
```
Comment (vil02) [inline src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java:36]: What do you think about moving it into some method with a signature like 
```java
private static void checkInput(final int weightCapacity, final int[] weights, final int[] values);
```

File: src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java [hunk 10]
```diff
@@ -1,38 +1,51 @@
...
+        if (weightCapacity < 0) {
+            throw new IllegalArgumentException("Weight capacity should not be negative.");
+        }
+        if (weights == null || values == null || weights.length != values.length) {
+            throw new IllegalArgumentException("Input arrays must not be null and must have the same length.");
+        }
+        for (int w : weights) {
+            if (w <= 0) {
+                throw new IllegalArgumentException("Input array should not contain non-positive weight(s).");
+            }
```
Comment (vil02) [inline src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java:35]:
```suggestion
        if (Arrays.stream(weights).anyMatch(w -> w <= 0)) {
            throw new IllegalArgumentException("Input array should not contain non-positive weight(s).");
        }
```

This change requires `import java.util.Arrays;`.

File: src/test/java/com/thealgorithms/dynamicprogramming/KnapsackTest.java [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    public void testKnapSackThrowsForNegativeCapacity() {
+        int[] weights = {2, 3, 4, 5};
+        int[] values = {3, 4, 5, 6};
+        int weightCapacity = -5;
+        assertThrows(IllegalArgumentException.class, () -> { Knapsack.knapSack(weightCapacity, weights, values); });
+    }
+
+    @Test
+    public void testKnapSackThrowsForNegativeWeight() {
+        int[] weights = {2, -3, 4};
```
Comment (vil02) [inline src/test/java/com/thealgorithms/dynamicprogramming/KnapsackTest.java:76]:
```suggestion
        int[] weights = {2, 0, 4};
```

File: src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java [hunk 12]
```diff
@@ -1,38 +1,51 @@
...
-        for (i = 0; i <= n; i++) {
-            for (w = 0; w <= W; w++) {
-                if (i == 0 || w == 0) {
-                    rv[i][w] = 0;
-                } else if (wt[i - 1] <= w) {
-                    rv[i][w] = Math.max(val[i - 1] + rv[i - 1][w - wt[i - 1]], rv[i - 1][w]);
-                } else {
-                    rv[i][w] = rv[i - 1][w];
+        // DP table to store the state of the maximum possible return for a given weight capacity.
+        int[] dp = new int[weightCapacity + 1];
```
Comment (vil02) [inline src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java:39]: If you insist on having this comment, why not to create a method like `private static int[] initializeDpArray(final int weightCapacity)`;

If you decide to have it, please check here if `weightCapacity >= 0` (not in the method `checkInput` mentioned earlier).

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java [hunk 13]
```diff
@@ -1,38 +1,50 @@
...
+ *
+ * @see <a href="https://en.wikipedia.org/?title=0-1_Knapsack_problem">0-1 Knapsack Problem </a>
  */
-public class Knapsack {
+public final class Knapsack {
 
-    private static int knapSack(int W, int[] wt, int[] val, int n) throws IllegalArgumentException {
-        if (wt == null || val == null) {
-            throw new IllegalArgumentException();
+    private Knapsack(){}
```
Comment (vil02) [inline src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java:13]:
```suggestion
    private Knapsack(){
    }
```

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java [hunk 14]
```diff
@@ -1,38 +1,51 @@
...
+ * @see <a href="https://en.wikipedia.org/?title=0-1_Knapsack_problem">0-1 Knapsack Problem </a>
  */
-public class Knapsack {
+public final class Knapsack {
 
-    private static int knapSack(int W, int[] wt, int[] val, int n) throws IllegalArgumentException {
-        if (wt == null || val == null) {
-            throw new IllegalArgumentException();
+    private Knapsack(){
+    }
```
Comment (vil02) [inline src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java:14]:
```suggestion
    private Knapsack() {
    }
```

**Review by vil02 — CHANGES_REQUESTED** [review]

**Review by BamaCharanChhandogi — DISMISSED** [review]

File: src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java [hunk 15]
```diff
@@ -1,38 +1,56 @@
...
-public class Knapsack {
+public final class Knapsack {
+
+    private Knapsack() {
+    }
 
-    private static int knapSack(int W, int[] wt, int[] val, int n) throws IllegalArgumentException {
-        if (wt == null || val == null) {
-            throw new IllegalArgumentException();
+    // Validate input for knapsack problem, otherwise throws exception
```
Comment (vil02) [inline src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java:18]:
```suggestion
```
I am not sure if this comment is needed. If you insist on having it, please use _doxygen_-like comment.

File: src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java (hunk 12 above)
Comment (Manan-09) [reply src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java:44]: Pardon @vil02  , which comment you are talking about?

Comment (Manan-09) [reply src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java:44]: Sorry @vil02, looks like already replied to your comment but it was in Pending state. Please check my [here](https://github.com/TheAlgorithms/Java/pull/4368#discussion_r1328132121)
Please tell if any changes needed

Comment (vil02) [reply src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java:43]: > I belive creating other method initializeDpArray will be overkill,

Understood.

> and why can't we check weightCapacity < 0 in our validation method?

If the method `initializeDpArray` would be introduced, I would suggest to make the `weightCapacity < 0` there.

**Review by vil02 — APPROVED** [review]
